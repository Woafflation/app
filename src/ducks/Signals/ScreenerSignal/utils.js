export const FREQUENCY_REAL_TIME = { label: 'Real-time', cooldown: '5m' }

export const SCREENER_DEFAULT_SIGNAL = {
  settings: {
    type: 'screener_signal',
    metric: 'social_volume_total',
    channel: 'telegram',
    operation: {
      selector: {}
    }
  },
  isActive: true,
  description: 'Notify me about changes inside the monitoring Screener',
  cooldown: FREQUENCY_REAL_TIME.cooldown
}

export const SCREENER_FREQUENCES = [
  FREQUENCY_REAL_TIME,
  { label: 'Once a day', cooldown: '1d' },
  { label: 'Once a week', cooldown: '7d' }
]
