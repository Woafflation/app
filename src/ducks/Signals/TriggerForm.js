import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import cx from 'classnames'
import { compose } from 'recompose'
import { graphql } from 'react-apollo'
import {
  Button,
  Input,
  Select,
  Checkboxes,
  Selector,
  SearchWithSuggestions,
  Message
} from '@santiment-network/ui'
import { Formik, Form, Field } from 'formik'
import { connect } from 'react-redux'
import { selectIsTelegramConnected } from './../../pages/UserSelectors'
import { allProjectsForSearchGQL } from './../../pages/Projects/allProjectsGQL'
import styles from './TriggerForm.module.scss'

const METRICS = [
  { label: 'Price', value: 'price' },
  { label: 'Trending Words', value: 'trendingWords' }
]

const OPTIONS = {
  price: [
    { label: '% Threshold', value: 'price_percent_change' },
    { label: 'Absolute', value: 'price_absolute_change' }
  ]
}

class TriggerForm extends React.Component {
  state = {
    metric: 'price',
    target: 'santiment',
    option: 'price_percent_change',
    cooldown: '24h',
    timeWindow: '24h',
    channels: this.props.isTelegramConnected ? ['Telegram'] : []
  }

  constructor (props) {
    super(props)
    this.form = React.createRef()
  }

  static propTypes = {
    onSettingsChange: PropTypes.func.isRequired,
    isTelegramConnected: PropTypes.bool.isRequired
  }

  static defaultProps = {
    isTelegramConnected: false
  }

  render () {
    const {
      data: { allProjects = [] }
    } = this.props
    return (
      <Formik
        initialValues={{
          percentThreshold: 5
        }}
        isInitialValid
        validate={values => {
          let errors = {}
          if (!values.percentThreshold) {
            errors.percentThreshold = 'Required'
          } else if (values.percentThreshold <= 0) {
            errors.percentThreshold = 'Must be more 0'
          }
          if (this.state.channels[0] !== 'Telegram') {
            errors.channels = 'You must setup notification channel'
          }
          return errors
        }}
        ref={this.form}
        onSubmit={(values, { setSubmitting }) => {
          this.props.onSettingsChange({ values, ...this.state })
        }}
      >
        {({
          values,
          errors,
          touched,
          isSubmitting,
          handleChange,
          handleBlur,
          isValid,
          ...rest
        }) => (
          <Form>
            <div className={styles.row}>
              <label>Metrics</label>
            </div>
            <div className={styles.row}>
              <div className={styles.Field}>
                <Select
                  placeholder='Choose a metric'
                  options={METRICS}
                  disabled
                  onChange={data => this.handleChange('metric', data)}
                  value={this.state.metric}
                />
              </div>
              {this.state.metric !== 'trendingWords' && (
                <div className={styles.Field}>
                  <Select
                    placeholder='Choose an option'
                    disabled
                    options={OPTIONS[this.state.metric]}
                    onChange={data => this.handleChange('option', data)}
                    value={this.state.option}
                  />
                </div>
              )}
            </div>
            <div className={styles.row}>
              <div className={styles.Field}>
                <label>Asset</label>
                <Select
                  placeholder='For example, ethereum...'
                  options={allProjects.map(asset => ({
                    label: asset.slug,
                    value: asset.slug
                  }))}
                  onChange={data => this.handleChange('target', data)}
                  value={this.state.target}
                />
              </div>
            </div>
            {this.state.metric !== 'trendingWords' && (
              <div className={styles.row}>
                <div className={styles.Field}>
                  <label>Threshold</label>
                  <Field
                    value={values.percentThreshold}
                    id='percentThreshold'
                    autoComplete='nope'
                    type='number'
                    name='percentThreshold'
                    placeholder='Setup the threshold'
                    isError={errors.percentThreshold}
                    defaultValue={errors.percentThreshold}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    component={Input}
                  />
                </div>
                <div className={styles.Field}>
                  <label>Time Window</label>
                  <Input
                    value={this.state.timeWindow}
                    id='timeWindow'
                    autoComplete='nope'
                    type='text'
                    disabled
                    name='timeWindow'
                    placeholder='setup the time window'
                    onChange={this.handleInputChange}
                  />
                </div>
              </div>
            )}
            <div className={styles.row}>
              <label>Notification settings</label>
            </div>
            <div className={styles.row}>
              <div className={styles.Field}>
                <label>Cooldown</label>
                <Selector
                  id='cooldown'
                  options={['1h', '24h']}
                  onSelectOption={this.handleSelectCooldown}
                  defaultSelected={this.state.cooldown}
                />
              </div>
            </div>
            <div className={styles.row}>
              <Checkboxes
                options={['Email', 'Telegram']}
                disabledIndexes={
                  this.props.isTelegramConnected
                    ? ['Email']
                    : ['Email', 'Telegram']
                }
                defaultSelectedIndexes={this.state.channels}
                onSelect={this.handleNotificationChannels}
                style={{ marginRight: '15px' }}
              />
              {!this.props.isTelegramConnected && (
                <Button
                  className={styles.connectLink}
                  variant='ghost'
                  as={Link}
                  to='/account'
                >
                  Telegram<span className={styles.connectLink}>Connect</span>
                </Button>
              )}
            </div>

            {errors.channels && (
              <div className={cx(styles.row, styles.messages)}>
                <Message variant='warn'>{errors.channels}</Message>
              </div>
            )}
            <Button
              type='submit'
              disabled={!isValid || isSubmitting}
              isActive={isValid && !isSubmitting}
              variant={'fill'}
              accent='positive'
            >
              Continue
            </Button>
          </Form>
        )}
      </Formik>
    )
  }

  handleSelectCooldown = cooldown => {
    this.setState({ cooldown })
  }

  handleChange = (field, data = {}) => {
    this.setState(prevState => {
      if (field === 'metric') {
        return {
          metric: data.value,
          option: prevState.metric !== data.value ? null : prevState.option
        }
      }
      return {
        [field]: (data || {}).value
      }
    })
  }

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value })
  }

  handleNotificationChannels = (channel, { selectedIndexes }) => {
    this.setState({ channels: selectedIndexes }, () => {
      this.form.current.validateForm()
    })
  }
}

const mapStateToProps = state => {
  return {
    isTelegramConnected: selectIsTelegramConnected(state)
  }
}

const enhance = compose(
  connect(mapStateToProps),
  graphql(allProjectsForSearchGQL)
)

export default enhance(TriggerForm)
