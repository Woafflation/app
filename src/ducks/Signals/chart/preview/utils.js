export const mapWithTimeseries = items =>
  items.map(item => ({
    ...item,
    datetime: +new Date(item.datetime),
    index: item.datetime
  }))

export const mapWithMidnightTime = date => {
  const datetime = new Date(date)
  datetime.setUTCHours(0, 0, 0, 0)
  return +new Date(datetime)
}

export const mapWithTimeseriesAndYCoord = (
  triggered,
  { key, dataKey, historicalTriggersDataKey = dataKey },
  timeseries,
  toDayConversion = true
) => {
  return triggered.map(point => {
    const date = toDayConversion
      ? mapWithMidnightTime(point.datetime)
      : +new Date(point.datetime)
    const item = timeseries.find(({ datetime }) => datetime === date)

    const yCoord =
      item && item[dataKey] ? item[dataKey] : point[historicalTriggersDataKey]

    return { date, yCoord, ...point }
  })
}

export const cleanByDatakeys = (timeseries, dataKey) => {
  return timeseries.filter(item => item[dataKey] !== undefined)
}

export const mapToRequestedMetrics = (
  metrics,
  { interval, slug, from, to, timeRange, address }
) =>
  metrics.map(({ key, alias: name = key, reqMeta }) => ({
    name,
    slug,
    from,
    to,
    timeRange,
    interval,
    address,
    ...reqMeta
  }))
