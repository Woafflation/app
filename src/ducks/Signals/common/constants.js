export const SIGNAL_ROUTES = {
  MY_SIGNALS: '/sonar/feed/my-signals'
}
