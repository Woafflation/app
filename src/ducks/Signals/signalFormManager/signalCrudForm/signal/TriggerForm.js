import React, { useState, useEffect } from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { compose } from 'recompose'
import { Formik, Form } from 'formik'
import { connect } from 'react-redux'
import isEqual from 'lodash.isequal'
import { selectIsTelegramConnected } from '../../../../../pages/UserSelectors'
import {
  fetchHistorySignalPoints,
  removeTrigger
} from '../../../common/actions'
import FormikCheckboxes from '../../../../../components/formik-santiment-ui/FormikCheckboxes'
import FormikEffect from '../../../../../components/formik-santiment-ui/FormikEffect'
import Button from '@santiment-network/ui/Button'
import { Checkbox } from '@santiment-network/ui'
import Message from '@santiment-network/ui/Message'
import {
  PRICE_PERCENT_CHANGE,
  METRIC_DEFAULT_VALUES,
  DEFAULT_FORM_META_SETTINGS,
  METRIC_TO_TYPES,
  TRENDING_WORDS
} from '../../../utils/constants'
import {
  couldShowChart,
  mapFormPropsToTrigger,
  validateTriggerForm
} from '../../../utils/utils'
import { TriggerFormMetricValues } from '../formParts/TriggerFormMetricValues'
import { TriggerFormMetricTypes } from '../formParts/TriggerFormMetricTypes'
import { TriggerFormFrequency } from '../formParts/TriggerFormFrequency'
import SignalPreview from '../../../chart/SignalPreview'
import Label from '@santiment-network/ui/Label'
import FormikSelect from '../../../../../components/formik-santiment-ui/FormikSelect'
import TriggerMetricTypesResolver from '../formParts/TriggerMetricTypesResolver'
import styles from './TriggerForm.module.scss'

const propTypes = {
  onSettingsChange: PropTypes.func.isRequired,
  isTelegramConnected: PropTypes.bool.isRequired,
  lastPriceItem: PropTypes.any,
  canRedirect: PropTypes.bool,
  settings: PropTypes.any,
  metaFormSettings: PropTypes.any,
  triggerMeta: PropTypes.any
}

export const TriggerForm = ({
  onSettingsChange,
  getSignalBacktestingPoints,
  isTelegramConnected = false,
  lastPriceItem,
  settings,
  metaFormSettings,
  trigger,
  removeSignal,
  onRemovedSignal
}) => {
  const formMetric =
    metaFormSettings && metaFormSettings.metric
      ? metaFormSettings.metric.value.value
      : PRICE_PERCENT_CHANGE

  metaFormSettings = { ...DEFAULT_FORM_META_SETTINGS, ...metaFormSettings }
  settings = {
    ...METRIC_DEFAULT_VALUES[formMetric],
    target: metaFormSettings.target.value
      ? metaFormSettings.target.value
      : settings.target,
    metric: metaFormSettings.metric.value
      ? metaFormSettings.metric.value
      : settings.metric,
    type: metaFormSettings.type.value
      ? metaFormSettings.type.value
      : settings.type,
    signalType: metaFormSettings.signalType.value
      ? metaFormSettings.signalType.value
      : settings.signalType,
    ethAddress: metaFormSettings.ethAddress,
    ...settings
  }

  const [initialValues, setInitialValues] = useState(settings)

  useEffect(() => {
    couldShowChart(initialValues.metric) &&
      getSignalBacktestingPoints(initialValues)
  }, [])

  const defaultType = metaFormSettings.type

  return (
    <Formik
      initialValues={initialValues}
      isInitialValid
      enableReinitialize
      validate={validateTriggerForm}
      onSubmit={values => {
        onSettingsChange(values)
      }}
    >
      {({
        values,
        errors,
        isSubmitting,
        handleChange,
        handleBlur,
        setFieldValue,
        isValid,
        validateForm
      }) => {
        const {
          metric,
          type,
          target,
          frequencyType,
          frequencyTimeType,
          isRepeating,
          ethAddress
        } = values
        const typeSelectors = METRIC_TO_TYPES[(metric || {}).value]

        const { price } = lastPriceItem || {}

        const showChart = couldShowChart(metric)
        const isTrendingWords = metric.value === TRENDING_WORDS

        return (
          <Form className={styles.TriggerForm}>
            <FormikEffect
              onChange={(current, prev) => {
                let { values: newValues } = current
                if (
                  !prev.values.metric ||
                  newValues.metric.value !== prev.values.metric.value ||
                  newValues.type.value !== prev.values.type.value
                ) {
                  const defaultValues =
                    METRIC_DEFAULT_VALUES[newValues.type.metric] || {}
                  newValues = {
                    ...defaultValues,
                    ...newValues
                  }
                  setInitialValues(newValues)
                  validateForm()
                }

                if (!isEqual(newValues, prev.values)) {
                  const lastErrors = validateTriggerForm(newValues)
                  const isError = Object.keys(newValues).some(
                    key => lastErrors[key]
                  )

                  const canLoadChart =
                    newValues && couldShowChart(newValues.metric)

                  newValues.target &&
                    !isError &&
                    canLoadChart &&
                    getSignalBacktestingPoints(newValues)
                }
              }}
            />

            <div className={styles.triggerFormItem}>
              <TriggerFormMetricTypes
                metaFormSettings={metaFormSettings}
                setFieldValue={setFieldValue}
                metric={metric}
              />

              <TriggerMetricTypesResolver
                isTrendingWords={isTrendingWords}
                address={ethAddress}
                metric={metric}
                target={target}
                metaFormSettings={metaFormSettings}
                setFieldValue={setFieldValue}
              />

              {!metric.hidden && typeSelectors && typeSelectors.length > 1 && (
                <div className={cx(styles.row)}>
                  <div className={cx(styles.Field, styles.fieldFilled)}>
                    <Label accent='waterloo' className={styles.label}>
                      Condition
                    </Label>
                    <FormikSelect
                      name='type'
                      isClearable={false}
                      isSearchable
                      isDisabled={defaultType.isDisabled}
                      defaultValue={defaultType.value}
                      placeholder='Choose a type'
                      options={typeSelectors}
                      isOptionDisabled={option => !option.value}
                    />
                  </div>
                </div>
              )}

              <TriggerFormMetricValues lastPrice={price} values={values} />

              {!isTrendingWords && (
                <TriggerFormFrequency
                  metaFormSettings={metaFormSettings}
                  setFieldValue={setFieldValue}
                  frequencyType={frequencyType}
                  metric={type.metric}
                  frequencyTimeType={frequencyTimeType}
                />
              )}

              {showChart && (
                <div className={cx(styles.row, styles.signalPreview)}>
                  <SignalPreview target={target.value} type={type.metric} />
                </div>
              )}

              <div className={cx(styles.row, styles.isRepeatingRow)}>
                <div className={styles.isRepeating}>
                  <Checkbox
                    isActive={isRepeating}
                    name='isRepeating'
                    className={styles.repeatingItem}
                    onClick={() => {
                      setFieldValue('isRepeating', !isRepeating)
                    }}
                  />
                  <span
                    className={styles.repeatingItem}
                    onClick={() => {
                      setFieldValue('isRepeating', !isRepeating)
                    }}
                  >
                    {isRepeating ? 'Task never ends' : 'Task fires only once'}
                  </span>
                </div>
              </div>

              <div className={styles.row}>
                <div className={styles.Field}>
                  <Label accent='waterloo' className={styles.label}>
                    Notify me via
                  </Label>
                  <div className={styles.notifyBlock}>
                    <FormikCheckboxes
                      name='channels'
                      labelOnRight
                      disabledIndexes={['Email']}
                      options={['Email', 'Telegram']}
                    />
                    {!isTelegramConnected && (
                      <Button
                        className={styles.connectLink}
                        variant='ghost'
                        as={Link}
                        to='/account'
                      >
                        <span className={styles.connectLink}>Connect</span>
                      </Button>
                    )}
                  </div>
                </div>
              </div>

              {errors.channels && (
                <div className={cx(styles.row, styles.messages)}>
                  <Message variant='warn'>{errors.channels}</Message>
                </div>
              )}
            </div>

            <div className={styles.controls}>
              <Button
                type='submit'
                disabled={!isValid || isSubmitting}
                isActive={isValid && !isSubmitting}
                variant={'fill'}
                accent='positive'
              >
                Continue
              </Button>
            </div>
          </Form>
        )
      }}
    </Formik>
  )
}

TriggerForm.propTypes = propTypes

const mapStateToProps = state => {
  return {
    isTelegramConnected: selectIsTelegramConnected(state),
    lastPriceItem: state.signals.points
      ? state.signals.points[state.signals.points.length - 1]
      : undefined
  }
}

const mapDispatchToProps = dispatch => ({
  getSignalBacktestingPoints: payload => {
    dispatch(fetchHistorySignalPoints(mapFormPropsToTrigger(payload)))
  },
  removeSignal: id => {
    dispatch(removeTrigger(id))
  }
})

const enhance = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)

export default enhance(TriggerForm)
