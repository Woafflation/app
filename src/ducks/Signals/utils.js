const getTimeWindowUnit = timeWindow => {
  if (!timeWindow) return undefined
  const value = timeWindow.replace(/[0-9]/g, '')
  return {
    value,
    label: (() => {
      switch (value) {
        case 'd':
          return 'days'
        case 'm':
          return 'minutes'
        default:
          return 'hours'
      }
    })()
  }
}

const getTarget = target => {
  // TODO: only for one asset as target
  const { slug } = target
  return { value: slug, label: slug }
}

export const DAILY_ACTIVE_ADDRESSES = 'daily_active_addresses'
export const PRICE_PERCENT_CHANGE = 'price_percent_change'
export const PRICE_VOLUME_DIFFERENCE = 'price_volume_difference'

export const PRICE_CHANGE_TYPES = {
  MOVING_UP: 'percent_up',
  MOVING_DOWN: 'percent_down'
}
export const pricePercentChangeUp = {
  value: PRICE_PERCENT_CHANGE,
  label: 'Moving up %',
  type: PRICE_CHANGE_TYPES.MOVING_UP
}

const getType = type => {
  const ALL_TYPES = {
    price_percent_change: { ...pricePercentChangeUp },
    daily_active_addresses: {
      label: 'Daily Active Addresses'
    },
    price_volume_difference: {
      label: 'Price/volume difference'
    }
  }
  return { value: type, ...ALL_TYPES[type] }
}

const getTriggerOperation = (model, percentThreshold) => {
  const mapped = {}
  mapped[model.type] = percentThreshold
  return mapped
}

const getMetric = type => {
  if (type === 'price_percent_change') {
    // TODO: add absolute price changing
    return { label: 'Price', value: 'price' }
  }
  return getType(type)
}

export const mapTriggerToFormProps = currentTrigger => {
  if (!currentTrigger) {
    return undefined
  }
  const settings = currentTrigger.settings

  return {
    cooldown: currentTrigger.cooldown,
    isRepeating: currentTrigger.isRepeating,
    isActive: currentTrigger.isActive,
    isPublic: currentTrigger.isPublic,
    metric: getMetric(settings.type),
    timeWindow: settings.time_window
      ? +settings.time_window.match(/\d+/)[0]
      : undefined,
    timeWindowUnit: settings.time_window
      ? getTimeWindowUnit(settings.time_window)
      : undefined,
    target: getTarget(settings.target),
    type: getType(settings.type),
    percentThreshold: settings.percent_threshold
      ? settings.percent_threshold
      : undefined,
    threshold: settings.threshold ? settings.threshold : undefined,
    channels: [settings.channel]
  }
}

export const mapFormPropsToTrigger = (formProps, prevTrigger) => {
  return {
    ...prevTrigger,
    settings: {
      channel: formProps.channels[0],
      percent_threshold: formProps.percentThreshold
        ? formProps.percentThreshold
        : undefined,
      threshold: formProps.threshold ? formProps.threshold : undefined,
      target: { slug: formProps.target.value },
      time_window: formProps.timeWindow
        ? formProps.timeWindow + '' + formProps.timeWindowUnit.value
        : undefined,
      type: getType(formProps.type.value).value,
      operation: getTriggerOperation(formProps.type, formProps.percentThreshold)
    },
    isPublic: !!formProps.isPublic,
    isRepeating: !!formProps.isRepeating,
    cooldown: formProps.cooldown ? formProps.cooldown : undefined,
    isActive: !!formProps.isActive
  }
}
export const mapValuesToTriggerProps = ({
  type,
  timeWindowUnit,
  timeWindow,
  percentThreshold,
  target,
  metric,
  threshold,
  cooldown
}) => ({
  cooldown,
  settings: (() => {
    const metricType = type ? type.value : undefined
    const time = timeWindowUnit ? timeWindow + timeWindowUnit.value : undefined

    const slug = { slug: target.value }

    const defaultValue = {
      target: slug,
      type: metricType,
      percent_threshold: percentThreshold,
      time_window: time
    }

    if (!metric) {
      return defaultValue
    }

    switch (metric.value) {
      case DAILY_ACTIVE_ADDRESSES:
        return {
          target: slug,
          type: metricType,
          time_window: time,
          percent_threshold: percentThreshold
        }
      case PRICE_VOLUME_DIFFERENCE:
        return {
          target: slug,
          type: metricType,
          threshold: threshold
        }
      default:
        return defaultValue
    }
  })()
})

export const METRICS = [
  { label: 'Price', value: 'price' },
  // { label: 'Trending Words', value: 'trendingWords' },
  { label: 'Daily Active Addresses', value: DAILY_ACTIVE_ADDRESSES },
  { label: 'Price/volume difference', value: PRICE_VOLUME_DIFFERENCE }
]

export const PRICE_TYPES = {
  price: [
    {
      label: 'Price changing',
      options: [
        {
          value: '',
          label: 'More than'
        },
        {
          value: '',
          label: 'Less than'
        },
        {
          value: '',
          label: 'Entering channel'
        },
        {
          value: '',
          label: 'Outside channel'
        }
      ]
    },
    {
      label: 'Percent change',
      options: [
        pricePercentChangeUp,
        {
          value: PRICE_PERCENT_CHANGE,
          label: 'Moving down %',
          type: PRICE_CHANGE_TYPES.MOVING_DOWN
        }
      ]
    }
  ],
  daily_active_addresses: [
    { label: 'Daily Active Addresses', value: DAILY_ACTIVE_ADDRESSES }
  ],
  price_volume_difference: [
    { label: 'Price/volume difference', value: PRICE_VOLUME_DIFFERENCE }
  ],
  price_percent_change: [
    { label: 'Price percentage change', value: PRICE_PERCENT_CHANGE }
  ]
}

export const ARGS = {
  price_volume_difference: ['threshold'],
  daily_active_addresses: ['percentThreshold', 'timeWindow'],
  price_percent_change: ['percentThreshold', 'timeWindow']
}

export const METRIC_DEFAULT_VALUES = {
  price_percent_change: {
    cooldown: '24h',
    percentThreshold: 5,
    timeWindow: 24,
    timeWindowUnit: { label: 'hours', value: 'h' },
    type: pricePercentChangeUp,
    isRepeating: true,
    channels: ['telegram']
  },
  daily_active_addresses: {
    cooldown: '24h',
    percentThreshold: 200,
    timeWindow: 2,
    timeWindowUnit: { label: 'days', value: 'd' },
    type: { label: 'Daily Active Addresses', value: DAILY_ACTIVE_ADDRESSES },
    isRepeating: true,
    channels: ['telegram']
  },
  price_volume_difference: {
    cooldown: '24h',
    threshold: 0.002,
    type: {
      label: 'Price/volume difference',
      value: PRICE_VOLUME_DIFFERENCE
    },
    isRepeating: true,
    channels: ['telegram']
  }
}

export const getTypeByMetric = metric => {
  if (metric.value === 'price') {
    return pricePercentChangeUp
  } else {
    return PRICE_TYPES[metric.value][0]
  }
}

export const mapTriggerToProps = ({ data: { trigger, loading, error } }) => {
  if (!loading && !trigger) {
    return {
      trigger: {
        isError: false,
        isEmpty: true,
        trigger: null,
        isLoading: false
      }
    }
  }
  if (
    !loading &&
    !(trigger || {}).trigger.settings.target.hasOwnProperty('slug')
  ) {
    return {
      trigger: {
        isError: true,
        isLoading: false,
        trigger: null,
        errorMessage: 'This is the unsupported signal format'
      }
    }
  }
  return {
    trigger: {
      trigger: (trigger || {}).trigger,
      isLoading: loading,
      isError: !!error
    }
  }
}
