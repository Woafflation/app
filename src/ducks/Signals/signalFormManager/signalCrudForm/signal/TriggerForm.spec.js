/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { TriggerForm } from './TriggerForm'
import { ETH_WALLET_METRIC } from '../../../utils/constants'

const allProjects = [
  {
    slug: 'santiment'
  }
]

describe('TriggerForm', () => {
  it('smoke', () => {
    const mockCb = jest.fn()
    const wrapper = shallow(
      <TriggerForm
        data={allProjects}
        canRedirect={true}
        isTelegramConnected={true}
        onSettingsChange={mockCb}
        getSignalBacktestingPoints={() => {}}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})

describe('TriggerForm EthWallet', () => {
  it('smoke', () => {
    const mockCb = jest.fn()

    const asset = 'nuga'
    const address = 'some_custom_address'

    const wrapper = shallow(
      <TriggerForm
        data={allProjects}
        canRedirect={true}
        isTelegramConnected={true}
        onSettingsChange={mockCb}
        getSignalBacktestingPoints={() => {}}
        metaFormSettings={{
          target: {
            value: {
              value: asset,
              label: asset
            }
          },
          metric: {
            value: { ...ETH_WALLET_METRIC }
          },
          type: {
            value: { ...ETH_WALLET_METRIC }
          },
          ethAddress: address
        }}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
