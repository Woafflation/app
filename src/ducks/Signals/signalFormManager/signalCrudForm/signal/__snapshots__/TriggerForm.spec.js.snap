// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerForm smoke 1`] = `
<Formik
  enableReinitialize={true}
  initialValues={
    Object {
      "absoluteBorderLeft": 50,
      "absoluteBorderRight": 75,
      "absoluteThreshold": 5,
      "channels": Array [
        "Telegram",
      ],
      "ethAddress": undefined,
      "frequencyTimeType": Object {
        "label": "Hours",
        "value": "h",
      },
      "frequencyTimeValue": Object {
        "label": 1,
        "value": 1,
      },
      "frequencyType": Object {
        "label": "Once Per",
        "value": "onceper",
      },
      "isRepeating": true,
      "metric": Object {
        "label": "Price",
        "value": "price",
      },
      "percentThreshold": 5,
      "signalType": Object {
        "label": "Assets",
        "value": "assets",
      },
      "target": Object {
        "label": "santiment",
        "value": "santiment",
      },
      "threshold": 0.002,
      "timeWindow": 24,
      "timeWindowUnit": Object {
        "label": "Hours",
        "value": "h",
      },
      "type": Object {
        "label": "Moving up %",
        "metric": "price_percent_change",
        "value": "percent_up",
      },
    }
  }
  isInitialValid={true}
  onSubmit={[Function]}
  validate={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
>
  <Component />
</Formik>
`;
